ruby-bin/profile-viewer

```diff
@@ -5,51 +5,132 @@ require "json"
-      response.body = File.binread(file)
-    else
-      response.status = 404
-      response['Content-Type'] = 'text/plain'
-      response.body = "Not Found"
+      response['Content-Type'] = 'text/html'
+      response.body = File.binread(File.join(App::HTTP_ROOT, "index.html"))
     end
   end
-end
 
-class GetIndex < WEBrick::HTTPServlet::AbstractServlet
-  def do_GET request, response
-    response.status = 200
-    response['Content-Type'] = 'text/html'
-    response.body = File.binread(File.join(HTTP_ROOT, "index.html"))
-  end
-end
+  class Profile < WEBrick::HTTPServlet::AbstractServlet
+    def initialize server, profile, parsed_profile
+      super(server)
+      @profile = profile
+      @parsed_profile = parsed_profile
+    end
 
-class Profile < WEBrick::HTTPServlet::AbstractServlet
-  def initialize server, profile, parsed_profile
-    super(server)
-    @profile = profile
-    @parsed_profile = parsed_profile
+    def do_GET request, response
+      response.status = 200
+      response['Content-Type'] = 'text/html'
+      response.body = JSON.dump(@parsed_profile)
+    end
   end
 
-  def do_GET request, response
-    response.status = 200
-    response['Content-Type'] = 'text/html'
-    response.body = JSON.dump(@parsed_profile)
+  class << self
+    def start
+      configure
+      server.start
+    end
+
+    def stop
+      server.shutdown
+    end
+
+    private
+
+    def server
+      @server ||= WEBrick::HTTPServer.new(
+        :Port => port,
+        :DocumentRoot => HTTP_ROOT,
+        :StartCallback => lambda {
+          should_get = "http://localhost:#{port}/from-url/" +
+            CGI.escape("http://localhost:#{port}/profile")
+          system("open #{should_get}")
+        }
+      )
+    end
+
+    def configure
+      unless PROFILE_PATH
+        $stderr.puts o
+        exit(2)
+      end
+
+      unless File.exist?(PROFILE_PATH)
+        $stderr.puts "no such file #{PROFILE_PATH}"
+        $stderr.puts
+        $stderr.puts o
+        exit(2)
+      end
+
+      server.mount "/from-url", GetIndex
+
+      if PROFILE_PATH
+        server.mount "/profile", Profile, full_path, parsed_file
+        server.mount "/source", Source, valid_files
+      end
+    end
+
+    def valid_files
+      @valid_files ||= Set.new.tap do |valid_files|
+        parsed_file["threads"].each do |thread|
+          thread["stringArray"].map! do |str|
+            if str.start_with?("/") && File.exist?(str)
+              valid_files << str
+              "http://localhost:#{port}/source#{str}"
+            else
+              str
+            end
+          end
+        end
+      end
+    end
+
+    def full_path
+      @full_path ||= File.expand_path PROFILE_PATH
+    end
+
+    def parsed_file
+      @parsed_file ||= JSON.parse File.read full_path
+    end
+
+    def port
+      @port ||= begin
+        port = START_PORT
+        loop do
+          break port unless system("lsof -i:#{port}", out: File::NULL)
```

Comment:
Instead of using `lsof` could we just try binding to the port and catching exceptions?  I'd rather not depend on a particular command line tool

Reply:
Done and done. Much more of a minimal change now.